connecting Panacea and Asclepius

c-brenn opened the pull request:
outline of how Panacea will depend on and interact with Asclepius.

c-brenn committed:
adds module to wrap all HTTP interactions with Asclepius

c-brenn committed:
asclepius dependent endpoints

changes router so `/api` routes are in their own scope as we will probably
change the pipeline they use.

adds a nest `/asclepius` scope for routes that will depend on asclepius being
available.

adds a Plug that requires Asclepius to be available before allowing requests
through.

c-brenn committed:
removes `async: true` from tests that share state

houli commented on panacea/test/plugs/require_asclepius_test.exs:
```diff
@@ -0,0 +1,27 @@
+    end
+  end
+
+  describe "whe asclepius is available" do
```
"when Asclepius"

houli commented on panacea/lib/panacea/asclepius/remote.ex:
```diff
@@ -0,0 +1,17 @@
+  def ping do
+    asclepius_uri("/ping")
+    |> HTTPoison.get([recv_timeout: @ping_timeout])
+    |> case() do
```
`case HTTPoison.get(asclepius_uri("/ping"), [recv_timeout: @ping_timeout]) do`
probably easier to read here

c-brenn committed:
mock Asclepius in test environment

`Panacea.Asclepius.Remote` is now a behaviour module - describes the contract
that any adapter for talking to Asclepius (or mocking it) needs to implement.

In production and dev environments, the adapter is
`Panacea.Asclepius.Remote.HTTP`. In test it's `Panacea.Asclepius.Remote.Mock`

c-brenn committed:
adds functions for fetching `/ddis`

22a commented:
Brilliant work, love the /asclepius pipeline! 👹 closes #48

c-brenn committed:
cleanup DDI fetching

c-brenn committed:
Merge branch 'iteration-2' into conor/chat-with-asclepius

c-brenn commented:
A bit going on here.

Adds a plug to handle requests that need Asclepius to be booted

Adds modules to abstract making requests to Asclepius -> Http and Mock
The module used is configured per environment - see `config :panacea, :asclepius_api`
This should make it very easy to test controllers that need to interact with asclepius.
A behaviour module is used to ensure Http and Mock implement the same contract

c-brenn committed:
Merge branch 'iteration-2' into conor/chat-with-asclepius

houli commented on panacea/lib/panacea/asclepius/remote/http.ex:
```diff
@@ -0,0 +1,40 @@
+  def ping do
+    asclepius_uri("/ping")
+    |> HTTPoison.get([recv_timeout: @ping_timeout])
+    |> case() do
```
I think this was nicer and easier to understand with just `case HTTPoison.get(asclepius_uri("/ping"), [recv_timeout: @ping_timeout])`

c-brenn replied:
Might change it to pipe into `decode_response` instead.
Then all of the `Remote` callbacks will have the same return type -> `{:ok, response} | {:error, reason}`.
Callers can then choose to ignore or use the response/error

houli replied:
👵 knows best

c-brenn committed:
all Remote callbacks have the same return type now
